### PR TITLE
Refactor callback registration to avoid NoEntitySpecifiedError

### PR DESCRIPTION
**Pull request by lkempf**
Currently the error `homeassistant.exceptions.NoEntitySpecifiedError: No entity id specified for entity None` gets raised on startup. This is due to the `_on_updated` callback being called before the entity has been initialized fully. To prevent this I used the [lifecycle hooks](https://developers.home-assistant.io/docs/core/entity#lifecycle-hooks) to register the update callback only after the entity has been added to HA.

### CODE REVIEW

**Comment by dbuezas** [comment]
Hey @lkempf,
Thanks for the PR!

There is also `update_before_add` which can be set to false https://github.com/dbuezas/eq3btsmart/blob/main/custom_components/dbuezas_eq3btsmart/climate.py#L112
The issue with that is that the component takes 5 minutes to get the first update.

What I used to do in the past is to trigger an update manually the first time the `available` getter is called, but that felt so hacky...
Your solution calling it on added is much better, can you combine it with setting update_before_add to false instead of adding the callback later?

**Comment by lkempf** [comment]
I'll test the proposed change. Nevertheless, I think that the callback should be subscribed when the entity is added and more importantly also unsubscribed when the entity is removed.

**Comment by dbuezas** [comment]
> I think that the callback should be subscribed when the entity is added
Then the first update will happen but be ignored

> unsubscribed when the entity is removed.
The updates don't happen unless the component requests them, so this is not that necessary. The only corner case is if there is a request in progress, in which case you are right (although this should be very rare)
WDYT about checking if the entity is still alive before calling `self.schedule_update_ha_state` in `def _on_updated:` instead?

**Comment by lkempf** [comment]
> Then the first update will happen but be ignored

I fixed this by checking whether the thermostat has already received an update and calling the update callback in `async_added_to_hass` if this is the case (https://github.com/lkempf/eq3btsmart/blob/ba7f47120fa4a3b903ee7c6523af6af07915a278/custom_components/dbuezas_eq3btsmart/climate.py#L157).
Doing as you suggested (`update_on_add=False` and starting the first update in `async_added_to_hass`) seems to behave the same including that startup won't finish until the update is done.

> The only corner case is if there is a request in progress, in which case you are right (although this should be very rare)

I agree that this doesn't sound all that bad.

> WDYT about checking if the entity is still alive before calling self.schedule_update_ha_state in def _on_updated: instead?

I don't know if there is a documented way to do that without using the hooks.

In conclusion I see three approaches to implement this:
- Set the callback in `__init__` and request the first update in `async_added_to_hass`.
- Set the callback using the hooks, don't update before add and request the first update in `async_added_to_hass`.
- Set the callback using the hooks, update before add and detect whether an update has been missed in `async_added_to_hass`.

**Comment by dbuezas** [comment]
I took the freedom of rebasing and committing what I mean to your PR as a suggestion.
What do you think about this variation?

File: custom_components/dbuezas_eq3btsmart/climate.py [hunk 1]
```diff
@@ -165,19 +165,10 @@ def __init__(self, _mac: str, _device_name: str, _hass: HomeAssistant):
...
-
-        # Small hack: HA might call async_update before the entity has been fully added.
-        # In this case the first update callback will be missed and _is_available won't be set.
-        # We circumvent this by checking whether the mode of the thermostat has been set.
-        # If this is the case the first update must have been sucessful.
-        if self._thermostat.mode != eq3.Mode.Unknown:
-            self._on_updated()
-
-        self._thermostat.on_update = self._on_updated
+        self.schedule_update_ha_state(force_refresh=True)
```
Comment (lkempf) [inline custom_components/dbuezas_eq3btsmart/climate.py:168]: I think this should be async_schedule_update_ha_state so that this doesn't block until the update is done.

**Comment by dbuezas** [comment]
> Doing as you suggested (update_on_add=False and starting the first update in async_added_to_hass) seems to behave the same including that startup won't finish until the update is done.

Oh, I didn't see your comment. Those delays finishing the startup bother me too :/ (and that's the only reason why there are no retries when updating state as there are when one sends a command)

**Comment by lkempf** [comment]
Apart from my comment this looks quite similar to what I tested and should work.

File: custom_components/dbuezas_eq3btsmart/climate.py (hunk 1 above)
Comment (dbuezas) [reply custom_components/dbuezas_eq3btsmart/climate.py:168]: Oh, you are right.

**Comment by dbuezas** [comment]
> should work.

but it blocks the finishing the ha core start, right?

**Comment by lkempf** [comment]
Yes that doesn't change and adding async_ doesn't help noticeably

**Comment by lkempf** [comment]
I also just tried `self.hass.async_create_task(self.async_update())` instead and this also doesn't help. It seems like the update task just get scheduled ahead of a task thats required for the startup to finish.

**Comment by dbuezas** [comment]
How annoying...
Another option (apart from your original idea) would be to trigger an update when the available property is queried:

```python
    async def async_available(self) -> bool:
        """Return if thermostat is available."""
        if self._thermostat.mode < 0:
                   self.async_schedule_update_ha_state(force_refresh=True)
        return True # so 
```

**Comment by lkempf** [comment]
I just noticed while testing, that bleak_retry_connector retries the connection more often than max_attempts (which defaults to 4). While somewhat off topic this seems to be one reason for the really long load times.

**Comment by dbuezas** [comment]
ok, `async_available` is not a thing, and updating on the synch function has the same issue.
I see available is checked during initialisation, so that's probably why it also delays startup.

I'll try doing it on some other getter

> I just noticed while testing, that bleak_retry_connector retries the connection more often than max_attempts (which defaults to 4). While somewhat off topic this seems to be one reason for the really long load times.

Yes, you are right. That's one of the reasons for which I use the retry_connector. On top of that I have my own retries because some times there are failures after connection. I have 7 Thermostats and 5 are on the edge of range, so without these retries they become unreliable

**Comment by dbuezas** [comment]
Or we could just not update right on startup, but that feels very wrong... One adds the thermostat sitting next to the bluetooth adaptor and then has to wait 5 minutes to see anything... :/

**Comment by dbuezas** [comment]
Ok, I found a way that doesn't block it and seems reasonable:

```python
    async def async_added_to_hass(self) -> None:
        _LOGGER.debug("[%s] adding", self._device_name)
        loop = asyncio.get_event_loop()
        loop.create_task(self._thermostat.async_update()) <--- doesn't block initialisation
```

combined with 
```python
async_add_entities(
        devices,
        update_before_add=False,  #<<--- therefore no component update blocks the startup
    )
```

**Comment by dbuezas** [comment]
Ugh... The firmware version is not shown in the device anymore (i assume because it is not known on startup)

**Comment by lkempf** [comment]
I don't think the firmware version is all that important and we can just remove it from DeviceInfo.

**Comment by dbuezas** [comment]
True.
I'll try this just in case:
```python
    async def fetch_serial(self):
        await self._thermostat.async_query_id()
        device_registry = dr.async_get(self.hass)
        device = device_registry.async_get_device(
            identifiers={(DOMAIN, self._mac)},
        )
        if device:
            device_registry.async_update_device(
                device_id=device.id, sw_version=self._thermostat.firmware_version
            )

```

**Comment by dbuezas** [comment]
image
works!

**Comment by dbuezas** [comment]
Thank you for your contribution and discussion @lkempf!

**Comment by lkempf** [comment]
Glad to have helped and thanks for maintaining the integration
